src/ucx_plugin.c

```diff
@@ -5,42 +5,580 @@
+    ucx_init_context(&workers[dev].ctx, dev);
+    ucx_init_worker(workers[dev].ctx, &workers[dev].worker);
+    workers[dev].last_tag = tag;
+  }
+
+  *ctx    = workers[dev].ctx;
+  *worker = workers[dev].worker;
+
+  if (newtag != NULL) {
+    workers[dev].last_tag += 1;
+    *newtag = workers[dev].last_tag;
+  }
+
+  ucp_worker_progress(*worker);
+  workers[dev].count++;
+#else
+  ucx_init_context(ctx, dev);
+  ucx_init_worker(*ctx, worker);
+  if (newtag != NULL) {
+    *newtag = tag;
+  }
+#endif
+  pthread_mutex_unlock(&nccl_ucx_lock);
+  return ncclSuccess;
+}
+
+static ncclResult_t nccl_ucx_free_worker(ucp_worker_h worker) {
+  int i;
+  int dummy;
+  struct ep_list *ep, *cur;
+
+  pthread_mutex_lock(&nccl_ucx_lock);
+  for(i = 0; i < ncclNIbDevs; i++) {
+    if (worker == workers[i].worker) {
+      workers[i].count--;
+      if (workers[i].count == 0){
+        ep = workers[i].eps;
+        while(ep){
+          cur = ep;
+          NCCLCHECK(socketReceive(ep->fd, &dummy, sizeof(int)));
+          ep = ep->next;
+          close(cur->fd);
+          free(cur);
+        }
+        ucp_worker_destroy(workers[i].worker);
+        ucp_cleanup(workers[i].ctx);
+        workers[i].eps    = NULL;
+        workers[i].worker = NULL;
+        workers[i].ctx    = NULL;
+      }
+      break;
+    }
+  }
+  pthread_mutex_unlock(&nccl_ucx_lock);
+
+  return ncclSuccess;
+}
+
+static ncclResult_t nccl_ucx_add_ep(ucp_worker_h worker, int fd) {
+  int i;
+
+  for(i = 0; i < ncclNIbDevs; i++) {
+    if (worker == workers[i].worker) {
+      struct ep_list *new_ep = (struct ep_list*)malloc(sizeof(struct ep_list));
+      new_ep->fd   = fd;
+      new_ep->next = workers[i].eps;
+      workers[i].eps = new_ep;
+      break;
+    }
+  }
+}
+
+static ncclResult_t ncclIbGetPciPath(char* devName, char** path, int* realPort) {
+  char devicePath[PATH_MAX];
+  snprintf(devicePath, PATH_MAX, "/sys/class/infiniband/%s/device", devName);
+  char* p = realpath(devicePath, NULL);
+  if (p == NULL) {
+    WARN("Could not find real path of %s", *devicePath);
+  } else {
+    // Merge multi-port NICs into the same PCI device
+    p[strlen(p)-1] = '0';
+    // And keep the real port aside (the ibv port is always 1 on recent cards)
+    *realPort = 0;
+    for (int d=0; d<ncclNIbDevs; d++) {
+      if (strcmp(p, ncclIbDevs[d].pciPath) == 0) (*realPort)++;
+    }
+  }
+  *path = p;
+  return ncclSuccess;
+}
+
+
+static int ibvWidths[] = { 1, 4, 8, 12 };
+static int ibvSpeeds[] = { 2500, 5000, 10000, 10000, 14000, 25000, 50000 };
+static int firstBitSet(int val, int max) {
+  int i = 0;
+  while (i<max && ((val & (1<<i)) == 0)) i++;
+  return i;
+}
+static int ncclIbWidth(int width) {
+  return ibvWidths[firstBitSet(width, sizeof(ibvWidths)/sizeof(int)-1)];
+}
+static int ncclIbSpeed(int speed) {
+  return ibvSpeeds[firstBitSet(speed, sizeof(ibvSpeeds)/sizeof(int)-1)];
+}
+
+uint64_t ncclParamSharpMaxComms();
+
+ncclResult_t nccl_ucx_init(ncclDebugLogger_t logFunction) {
```

Comment:
this func is same as ib plugin ncclIbInit() right?

Reply:
Right now it's almost the same as ncclIbInit, but still there some differences. I don't want to merge ncclIbInit and nccl_ucx_init because of the following:
1. UCX plugin currently supports only ib and roce but nothing stops us to add other transports (for instance tcp), it will require different init functions for ib and ucx as well as other helper functions (pciPath, getPropterties ...)
2. Having one init function will result in a lot of extern variables everywhere